File: assets/data/update_graph.py
import json
import os

# Konfiguracja
JSON_PATH = "school_data.json"
LOG_PATH = "debug_log.txt"

def log(msg):
    with open(LOG_PATH, "a", encoding="utf-8") as f:
        f.write(msg + "\n")
    print(msg)

def update_school_data():
    if os.path.exists(LOG_PATH): os.remove(LOG_PATH)
    log(f"Rozpoczynam aktualizację: {JSON_PATH}")
    if not os.path.exists(JSON_PATH):
        log(f"Błąd: Plik nie istnieje w {os.getcwd()}")
        return

    try:
        with open(JSON_PATH, 'r', encoding='utf-8') as f:
            data = json.load(f)
        log("Wczytano JSON.")
    except Exception as e:
        log(f"Błąd odczytu: {e}")
        return

    nodes = data.get('nodes', [])
    edges = data.get('edges', [])
    
    # 1. Usuwanie starych krawędzi dla parteru (FLOOR_0)
    #    Zakładam, że krawędzie na parterze łączą węzły z floor_id="FLOOR_0".
    #    Aby być bezpiecznym, usunę tylko te krawędzie, które łączą dwa węzły z FLOOR_0.
    #    Przy okazji zbudujemy mapę id -> floor_id dla szybkiego sprawdzania.
    node_floor_map = {n['id']: n['floor_id'] for n in nodes}
    
    new_edges = []
    for edge in edges:
        u, v = edge['from'], edge['to']
        # Jeśli oba węzły są na FLOOR_0, pomijamy tę krawędź (zostanie zastąpiona nową logiką)
        # UWAGA: Musimy zachować krawędzie między piętrami (np. schody), więc sprawdzamy oba.
        if node_floor_map.get(u) == "FLOOR_0" and node_floor_map.get(v) == "FLOOR_0":
            continue
        new_edges.append(edge)
    
    # 2. Dodawanie nowych węzłów (punkty pośrednie)
    #    Współrzędne są przybliżone (np. przesunięcie o 10-20px względem "rodzica" lub interpolacja).
    #    Dla uproszczenia, w tym skrypcie dodamy je z przykładowymi współrzędnymi,
    #    zakładając, że użytkownik poprawi je wizualnie w edytorze lub JSONie.
    #    Będę starał się "strzelać" w pobliże na podstawie nazw logicznych.
    
    new_nodes_definitions = [
        {"id": "PUNKT_1A", "floor_id": "FLOOR_0", "x": 700, "y": 164, "type": "CORRIDOR", "name": "Punkt 1a"}, # Blisko 1 (779, 164) w stronę m1/m2
        {"id": "PUNKT_1B", "floor_id": "FLOOR_0", "x": 850, "y": 164, "type": "CORRIDOR", "name": "Punkt 1b"}, # Blisko 1
        {"id": "PUNKT_4A", "floor_id": "FLOOR_0", "x": 750, "y": 325, "type": "CORRIDOR", "name": "Punkt 4a"}, # Blisko 4
        {"id": "PUNKT_4B", "floor_id": "FLOOR_0", "x": 880, "y": 325, "type": "CORRIDOR", "name": "Punkt 4b"},
        {"id": "PUNKT_4C", "floor_id": "FLOOR_0", "x": 810, "y": 380, "type": "CORRIDOR", "name": "Punkt 4c"},
        {"id": "PUNKT_4D", "floor_id": "FLOOR_0", "x": 810, "y": 420, "type": "CORRIDOR", "name": "Punkt 4d"},
        {"id": "PUNKT_6A", "floor_id": "FLOOR_0", "x": 750, "y": 600, "type": "CORRIDOR", "name": "Punkt 6a"},
        {"id": "PUNKT_11A", "floor_id": "FLOOR_0", "x": 500, "y": 840, "type": "CORRIDOR", "name": "Punkt 11a"},
        {"id": "PUNKT_11B", "floor_id": "FLOOR_0", "x": 600, "y": 800, "type": "CORRIDOR", "name": "Punkt 11b"},
        {"id": "PUNKT_12A", "floor_id": "FLOOR_0", "x": 426, "y": 900, "type": "CORRIDOR", "name": "Punkt 12a"},
        {"id": "PUNKT_12B", "floor_id": "FLOOR_0", "x": 380, "y": 838, "type": "CORRIDOR", "name": "Punkt 12b"},
        {"id": "PUNKT_12C", "floor_id": "FLOOR_0", "x": 400, "y": 920, "type": "CORRIDOR", "name": "Punkt 12c"},
        {"id": "PUNKT_12E", "floor_id": "FLOOR_0", "x": 100, "y": 830, "type": "CORRIDOR", "name": "Punkt 12e"}, # Blisko 13
    ]
    
    existing_node_ids = {n['id'] for n in nodes}
    for new_node in new_nodes_definitions:
        if new_node['id'] not in existing_node_ids:
            nodes.append(new_node)
            existing_node_ids.add(new_node['id']) # Zapobiega duplikatom przy ponownym uruchomieniu
            
    # 3. Definiowanie nowych krawędzi (ścieżek)
    #    Format: (from, to, weight) - waga domyślna np. 50 (wymaga przeliczenia odległości euklidesowej, ale tu uproszczę).
    #    Lepiej byłoby liczyć odległość, ale dla uproszczenia damy stałą lub "1", bo A* i tak użyje wag.
    #    Tutaj użyjemy funkcji pomocniczej do dodania krawędzi w obu kierunkach (chyba że graf skierowany?).
    #    Zakładam graf nieskierowany -> dodajemy jedną krawędź, a aplikacja (Flutter) obsługuje to dwukierunkowo?
    #    W oryginalnym pliku są krawędzie PUNKT_1 -> PUNKT_2 (skierowane?).
    #    Zazwyczaj w nawigacji są dwukierunkowe, ale w JSONie są "from" i "to". 
    #    Jeśli system wymaga dwukierunkowości jawnie, dodam w obie strony.
    #    Sprawdzając oryginał: PUNKT_1->PUNKT_2 jest, a PUNKT_2->PUNKT_1 nie widzę na szybko w wycinku.
    #    Przyjmę konwencję dodawania krawędzi "w jedną stronę" zgodnie z opisem użytkownika, 
    #    ale dla bezpieczeństwa nawigacji dodam też powrotne, jeśli to standard.
    #    Jednak trzymając się sciśle instrukcji "Z A możemy przejść do B", zrobię directed graph zgodnie z opisem.
    #    Jeśli nawigacja ma działać w obie strony, algorytm w Flutterze (A*) pewnie traktuje krawędzie jako nieskierowane
    #    lub wymaga par. Bez widoku kodu Dart ciężko ocenić, ale bezpieczniej jest dodać krawędzie.
    #    Zrobię na razie zgodnie z "Z X do Y", czyli skierowane.
    
    # Lista połączeń (from, to)
    connections = [
        ("PUNKT_1", "PUNKT_1A"), ("PUNKT_1", "PUNKT_1B"), ("PUNKT_1", "PUNKT_2"),
        ("PUNKT_2", "PUNKT_3"),
        ("PUNKT_3", "PUNKT_4"),
        ("PUNKT_4", "PUNKT_4A"), ("PUNKT_4", "PUNKT_4B"), ("PUNKT_4", "PUNKT_4C"),
        ("PUNKT_4C", "PUNKT_4D"),
        ("PUNKT_4D", "PUNKT_5"),
        ("PUNKT_5", "PUNKT_6"),
        ("PUNKT_6", "PUNKT_6A"),
        ("PUNKT_6A", "PUNKT_7"),
        ("PUNKT_7", "PUNKT_8"),
        ("PUNKT_8", "PUNKT_9"),
        ("PUNKT_9", "PUNKT_10"), ("PUNKT_9", "PUNKT_11"),
        ("PUNKT_11", "PUNKT_11A"), ("PUNKT_11", "PUNKT_11B"),
        ("PUNKT_11A", "PUNKT_12"),
        ("PUNKT_12", "PUNKT_12A"), ("PUNKT_12", "PUNKT_12B"),
        ("PUNKT_12A", "PUNKT_12C"),
        ("PUNKT_12B", "PUNKT_12C"),
        ("PUNKT_12B", "PUNKT_12E"), # Z Punkt_12b możemy przejść do Punkt_12e (zgodnie z korektą użytkownika)
        ("PUNKT_12E", "PUNKT_13")
    ]
    
    # Połączenia do sal (from, to)
    room_connections = [
        ("PUNKT_1A", "m1"), ("PUNKT_1A", "m2"),
        ("PUNKT_1", "m3"),
        ("PUNKT_1B", "m4"), ("PUNKT_1B", "DEG"), ("PUNKT_1B", "KUCHNIA"), # Zakładam KUCHNIA jako Wejście_kuchnia
        ("PUNKT_4A", "7"),
        ("PUNKT_4B", "8"), ("PUNKT_4B", "9"),
        ("PUNKT_4C", "K4"),
        ("PUNKT_4D", "K3"), ("PUNKT_4D", "SZATNIA_1"), # SZATNIA_1 lub SZATNIA_2
        ("PUNKT_7", "K1"), ("PUNKT_7", "K2"),
        ("PUNKT_8", "6"),
        ("PUNKT_10", "5"), ("PUNKT_10", "CENTRUM"), ("PUNKT_10", "BIBLIOTEKA"), ("PUNKT_10", "PEDAGOG"),
        ("PUNKT_11", "4b"), ("PUNKT_11", "4a"),
        ("PUNKT_11B", "STAIR_A_DOWN"), # Schody pierwsze dół
        ("PUNKT_11A", "3"),
        ("PUNKT_12", "2"),
        ("PUNKT_12B", "1"),
        ("PUNKT_12A", "B1"),
        ("PUNKT_12C", "B2"), ("PUNKT_12C", "B3"),
        ("PUNKT_12E", "GYM"), # Sala gimnastyczna 1
        ("PUNKT_13", "GYM_2"), # Sala gimnastyczna 2
        ("PUNKT_13", "PSYCHOLOG"),
        ("PUNKT_13", "STAIR_B_DOWN"), # Schody drugie dół
    ]
    
    all_connections = connections + room_connections
    
    # Dodajemy krawędzie
    for u, v in all_connections:
        # Sprawdzenie czy węzły istnieją w 'nodes' (lub starych lub nowych)
        # Ignorujemy błędy "missing node" w tym prostym skrypcie, 
        # ale w produkcji warto by rzucić warning.
        
        # Obliczenie wagi (dystans)
        node_u = next((n for n in nodes if n["id"] == u), None)
        node_v = next((n for n in nodes if n["id"] == v), None)
        
        weight = 100 # Default
        if node_u and node_v:
             weight = int(((node_u["x"] - node_v["x"])**2 + (node_u["y"] - node_v["y"])**2)**0.5)
        
        new_edges.append({
            "from": u,
            "to": v,
            "weight": weight,
            "type": "WALK"
        })
        
        # Opcjonalnie: dodaj krawędź powrotną v->u ?
        # Jeśli nawigacja ma działać w obie strony (np. wyjście z sali), to tak.
        # Użytkownik napisał "Z Punktu X możemy wejść do Y".
        # Ale zazwyczaj wychodzi się tą samą drogą.
        # ZOSTAWIMY dwukierunkowość, bo to standard w pathfindingu.
        new_edges.append({
            "from": v,
            "to": u,
            "weight": weight,
            "type": "WALK"
        })

    data['nodes'] = nodes
    data['edges'] = new_edges
    
    output_path = "school_data_updated.json"
    log(f"Zapisywanie do {output_path} ({len(nodes)} węzłów, {len(new_edges)} krawędzi)...")
    try:
        with open(output_path, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=4, ensure_ascii=False)
        log(f"Zapisano {output_path} pomyślnie.")
    except Exception as e:
        log(f"Błąd zapisu: {e}")

if __name__ == "__main__":
    update_school_data()

File: assets/data/update_graph.js
const fs = require('fs');
const path = require('path');

const JSON_PATH = path.join(__dirname, 'school_data.json');
const OUTPUT_PATH = path.join(__dirname, 'school_data.json'); // Overwrite directly

// GLOBAL OFFSET CONFIGURATION
const OFFSET_X = 0;
const OFFSET_Y = 0;

const NEW_NODES_RAW = [
    // --- GÓRNY KORYTARZ (Y=160) ---
    { id: "PUNKT_1A", x: 640, y: 160 },   // M2 (643) - OK
    { id: "PUNKT_M1", x: 686, y: 160 },   // [NEW] M1 (686)
    { id: "PUNKT_1", x: 765, y: 160 },    // M3 (762) - OK
    { id: "PUNKT_M4", x: 843, y: 160 },   // [MOVED] M4 (843)
    { id: "PUNKT_1B", x: 900, y: 160 },   // Kuchnia/Deg

    // --- PIONOWY ŁĄCZNIK ---
    { id: "PUNKT_2", x: 779, y: 219 },
    { id: "PUNKT_3", x: 805, y: 219 },
    { id: "PUNKT_4", x: 805, y: 325 },
    { id: "PUNKT_4A", x: 720, y: 325 },
    { id: "PUNKT_4B", x: 890, y: 325 },
    { id: "PUNKT_4C", x: 805, y: 367 },
    { id: "PUNKT_4D", x: 805, y: 415 },
    { id: "PUNKT_5", x: 805, y: 451 },
    { id: "PUNKT_6", x: 805, y: 544 },
    { id: "PUNKT_6A", x: 750, y: 544 },
    { id: "PUNKT_7", x: 750, y: 649 },
    { id: "PUNKT_8", x: 750, y: 753 },
    { id: "PUNKT_9", x: 750, y: 828 }, // Skrzyżowanie T
    { id: "PUNKT_10", x: 845, y: 845 },

    // --- DOLNY KORYTARZ (Y=828) ---
    // Kolejność od prawej (PUNKT_9) do lewej
    { id: "PUNKT_11", x: 632, y: 828 },    // [MOVED] Sala 4b (632)
    { id: "PUNKT_11B", x: 580, y: 828 },   // Sala 4a (581) / Schody A
    { id: "PUNKT_SAL3", x: 534, y: 828 },  // [NEW] Sala 3 (534)
    { id: "PUNKT_11A", x: 506, y: 828 },   // Punkt pomocniczy

    { id: "PUNKT_12", x: 440, y: 828 },      // Sala 2 (440) + Rozgałęzienie B
    { id: "PUNKT_SAL1", x: 340, y: 828 },    // [NEW] Sala 1 (340)
    { id: "PUNKT_12B", x: 319, y: 828 },     // Sala Gimnastyczna (Prawe)
    { id: "PUNKT_12E", x: 170, y: 828 },     // Długi odcinek
    { id: "PUNKT_13", x: 27, y: 828 },       // Sala Gimnastyczna (Lewe) / Schody / Psycholog

    // --- SEKCJA B (Y rośnie w dół, X=440) ---
    { id: "PUNKT_12A", x: 440, y: 880 },
    { id: "PUNKT_12C", x: 440, y: 940 },
];

// Apply offset
const NEW_NODES = NEW_NODES_RAW.map(n => ({
    ...n,
    x: n.x + OFFSET_X,
    y: n.y + OFFSET_Y
}));

const EDGES = [
    // Górny korytarz: 1A <-> M1 <-> 1 <-> M4 <-> 1B
    { from: "PUNKT_1A", to: "PUNKT_M1" },
    { from: "PUNKT_M1", to: "PUNKT_1" },
    { from: "PUNKT_1", to: "PUNKT_M4" },
    { from: "PUNKT_M4", to: "PUNKT_1B" },

    // Łącznik do głównego pionu
    { from: "PUNKT_1", to: "PUNKT_2" },

    // Pion
    { from: "PUNKT_2", to: "PUNKT_3" },
    { from: "PUNKT_3", to: "PUNKT_4" },
    { from: "PUNKT_4", to: "PUNKT_4A" },
    { from: "PUNKT_4", to: "PUNKT_4B" },
    { from: "PUNKT_4", to: "PUNKT_4C" },
    { from: "PUNKT_4C", to: "PUNKT_4D" },
    { from: "PUNKT_4D", to: "PUNKT_5" },
    { from: "PUNKT_5", to: "PUNKT_6" },
    { from: "PUNKT_6", to: "PUNKT_6A" },
    { from: "PUNKT_6A", to: "PUNKT_7" },
    { from: "PUNKT_7", to: "PUNKT_8" },
    { from: "PUNKT_8", to: "PUNKT_9" },
    { from: "PUNKT_9", to: "PUNKT_10" },

    // Dolny korytarz: 9 -> 11 -> 11B -> SAL3 -> 11A -> 12 -> SAL1 -> 12B -> 12E -> 13
    { from: "PUNKT_9", to: "PUNKT_11" },
    { from: "PUNKT_11", to: "PUNKT_11B" },
    { from: "PUNKT_11B", to: "PUNKT_SAL3" },
    { from: "PUNKT_SAL3", to: "PUNKT_11A" },
    { from: "PUNKT_11A", to: "PUNKT_12" },
    { from: "PUNKT_12", to: "PUNKT_SAL1" },
    { from: "PUNKT_SAL1", to: "PUNKT_12B" },
    { from: "PUNKT_12B", to: "PUNKT_12E" },
    { from: "PUNKT_12E", to: "PUNKT_13" },

    // Sekcja B
    { from: "PUNKT_12", to: "PUNKT_12A" },
    { from: "PUNKT_12A", to: "PUNKT_12C" },

    // Skrzyżowanie T na dole (Punkt 11 w pionie?)
    // Note: W oryginalnym grafie PUNKT_9 łączył się z 11. Teraz mamy łańcuch.
];

const ROOM_CONNECTIONS = [
    { room: "M1", node: "PUNKT_M1" }, // [NEW] Prostopadłe
    { room: "M2", node: "PUNKT_1A" }, // Blisko
    { room: "M3", node: "PUNKT_1" },  // Blisko
    { room: "M4", node: "PUNKT_M4" }, // [NEW] Prostopadłe

    { room: "Kuchnia", node: "PUNKT_1B" },
    { room: "Deg.", node: "PUNKT_1B" },
    { room: "7", node: "PUNKT_4A" },
    { room: "8", node: "PUNKT_4B" },
    { room: "9", node: "PUNKT_4B" },
    { room: "K4", node: "PUNKT_4C" },
    { room: "K3", node: "PUNKT_4D" },
    { room: "Szatnia", node: "PUNKT_4D" },
    { room: "K1", node: "PUNKT_7" },
    { room: "K2", node: "PUNKT_7" },
    { room: "6", node: "PUNKT_8" },
    { room: "5", node: "PUNKT_10" },
    { room: "Centrum", node: "PUNKT_10" },
    { room: "Biblioteka", node: "PUNKT_10" },
    { room: "Pedagog", node: "PUNKT_10" },

    { room: "4b", node: "PUNKT_11" },   // [NEW] Przesunięty węzeł
    { room: "4a", node: "PUNKT_11B" },
    { room: "Schody_pierwsz_dół", node: "PUNKT_11B" },
    { room: "3", node: "PUNKT_SAL3" },  // [NEW] Prostopadłe
    { room: "2", node: "PUNKT_12" },    // Idealnie (440)
    { room: "1", node: "PUNKT_SAL1" },  // [NEW] Prostopadłe

    { room: "B1", node: "PUNKT_12A" },
    { room: "B2", node: "PUNKT_12C" },
    { room: "B3", node: "PUNKT_12C" },

    { room: "Sala gimnastyczna", node: "PUNKT_12E" }, // Prawe wejście (W2 - przy Pielęgniarce)
    { room: "Sala gimnastyczna", node: "PUNKT_13" },
    { room: "Psycholog", node: "PUNKT_13" },
    { room: "Schody_drugie_dół", node: "PUNKT_13" }
];

function updateGraph() {
    try {
        const data = JSON.parse(fs.readFileSync(JSON_PATH, 'utf8'));

        // 1. Update/Add Nodes
        NEW_NODES.forEach(newNode => {
            const existingNodeIndex = data.nodes.findIndex(n => n.id === newNode.id);
            const nodeEntry = {
                id: newNode.id,
                floor_id: "FLOOR_0",
                x: newNode.x,
                y: newNode.y,
                type: "PATH_POINT",
                name: newNode.id.replace("PUNKT_", "Punkt ")
            };

            if (existingNodeIndex !== -1) {
                data.nodes[existingNodeIndex] = nodeEntry;
            } else {
                data.nodes.push(nodeEntry);
            }
        });

        // 2. Remove old edges for FLOOR_0 to rebuild them
        data.edges = data.edges.filter(e => {
            const n1 = data.nodes.find(n => n.id === e.from);
            const n2 = data.nodes.find(n => n.id === e.to);
            if ((n1 && n1.floor_id === 'FLOOR_0') || (n2 && n2.floor_id === 'FLOOR_0')) {
                return false;
            }
            return true;
        });

        // 3. Add new edges
        EDGES.forEach(edge => {
            const n1 = data.nodes.find(n => n.id === edge.from);
            const n2 = data.nodes.find(n => n.id === edge.to);
            if (n1 && n2) {
                const dist = Math.sqrt(Math.pow(n1.x - n2.x, 2) + Math.pow(n1.y - n2.y, 2));
                data.edges.push({
                    from: edge.from,
                    to: edge.to,
                    weight: Math.round(dist),
                    type: "WALK"
                });
            } else {
                console.warn(`Missing node for edge: ${edge.from} -> ${edge.to}`);
            }
        });

        // 4. Update Room Connections (Edges from Room to Node)
        ROOM_CONNECTIONS.forEach(conn => {
            let roomNode = data.nodes.find(n => n.name === conn.room || n.id === conn.room || (n.label && n.label.includes(conn.room)));

            if (!roomNode) {
                roomNode = data.nodes.find(n => n.name && n.name.toLowerCase().includes(conn.room.toLowerCase()) && n.floor_id === 'FLOOR_0');
            }

            if (roomNode) {
                const pathNode = data.nodes.find(n => n.id === conn.node);
                if (pathNode) {
                    const dist = Math.sqrt(Math.pow(roomNode.x - pathNode.x, 2) + Math.pow(roomNode.y - pathNode.y, 2));
                    data.edges.push({
                        from: roomNode.id,
                        to: pathNode.id,
                        weight: Math.round(dist),
                        type: "WALK"
                    });
                }
            } else {
                // Ignore missing rooms for now to keep it clean
            }
        });

        fs.writeFileSync(OUTPUT_PATH, JSON.stringify(data, null, 4));
        console.log(`Graph updated successfully with OFFSET X:${OFFSET_X} Y:${OFFSET_Y}`);

    } catch (e) {
        console.error("Error:", e);
    }
}

updateGraph();
